crates/snapbox/src/data.rs

```diff
@@ -145,6 +219,23 @@ impl Data {
                     }
                 }
                 DataInner::Text(data) => Self::text(data),
+                #[cfg(feature = "json")]
+                DataInner::Json(data) => match serde_json::to_string_pretty(&data) {
+                    Ok(text) => Self::text(text),
+                    Err(_) => Self::json(data),
+                },
+            },
+            #[cfg(feature = "json")]
+            DataFormat::Json => match self.inner {
+                DataInner::Binary(bin) => match serde_json::from_slice::<Value>(&bin) {
+                    Ok(json) => Self::json(json),
+                    Err(_) => Self::binary(bin),
+                },
+                DataInner::Text(text) => match serde_json::from_str::<Value>(&text) {
+                    Ok(json) => Self::json(json),
+                    Err(_) => Self::text(text),
+                },
+                DataInner::Json(json) => Self::json(json),
             },
         }
```

Comment:
I feel like there needs to be a better way to implement this before it explodes under the number of permutations we'll be making

Reply:
I debated using `match (format, self.inner)` but it still had problems with lots of permutations. I'm not quite sure how to make this better over all

Reply:
I think `match (self.inner, format)` will end up working the best

Something like
```rust
match (self.inner, format) {
    (Binary(inner), Binary) => Binary(inner),
    (Text(inner), Text) => Text(inner),
    (Json(inner), Json) => Json(inner),
    (Binary(inner), _) => {
        if is_binary(&data) => {
            Binary(inner)
        } else if Ok(inner) = String::from_utf8(inner) {
            Text(inner).try_coerce(format)
        } else {
            Binary(inner)
        }
    },
    (Text(inner), Json) => ...,
}
```

Then as we add different versions of structured data, we can just
```
(Json(inner), Toml) | (Yaml(inner), Toml) => Toml(inner),
```

Reply:
I'm not sure `|` would work since it would require complex `#[cfg(feature = "json")]`:
```
#[cfg(all(feature = "json", feature = "yaml", feature = "toml"))]
(Json(inner), Toml) | (Yaml(inner), Toml) => Toml(inner),
```